### ISSUE TITLE
Misleading webSocket() overloading

### ISSUE THREAD

**Issue by hazae41** [ktorio/ktor#1075]
### Ktor Version
1.1.4

### Ktor Engine Used(client or server and name)
Server: CIO
Client: CIO 

### JVM Version, Operating System and Relevant Context
Java 8
Windows 10

When I try to use WebSockets (ws) in a CIO server and a CIO client on the same machine it throws a 404 error

### Feedback
	20:08:53 [GRAVE] Exception in thread "DefaultDispatcher-worker-1"
	
	20:08:53 [GRAVE] io.ktor.client.features.BadResponseStatusException: Received bad status code: 404 Not Found. Expected status code < 300.
	
	20:08:53 [GRAVE]     at io.ktor.client.features.ExpectSuccess$Companion$install$1.invokeSuspend(ExpectSuccess.kt:22)
	
	20:08:53 [GRAVE]     at io.ktor.client.features.ExpectSuccess$Companion$install$1.invoke(ExpectSuccess.kt)
	
	20:08:53 [GRAVE]     at io.ktor.util.pipeline.SuspendFunctionGun.loop(PipelineContext.kt:278)
	
	20:08:53 [GRAVE]     at io.ktor.util.pipeline.SuspendFunctionGun.access$loop(PipelineContext.kt:63)
	
	20:08:53 [GRAVE]     at io.ktor.util.pipeline.SuspendFunctionGun.proceed(PipelineContext.kt:137)
	
	20:08:53 [GRAVE]     at io.ktor.util.pipeline.SuspendFunctionGun.execute(PipelineContext.kt:157)
	
	20:08:53 [GRAVE]     at io.ktor.util.pipeline.Pipeline.execute(Pipeline.kt:23)
	
	20:08:53 [GRAVE]     at io.ktor.client.call.HttpClientCall.receive(HttpClientCall.kt:55)
	
	20:08:53 [GRAVE]     at io.ktor.client.features.websocket.BuildersKt.webSocketRawSession(builders.kt:106)
	
	20:08:53 [GRAVE]     at io.ktor.client.features.websocket.BuildersKt$webSocketRawSession$1.invokeSuspend(builders.kt)
	
	20:08:53 [GRAVE]     at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
	
	20:08:53 [GRAVE]     at io.ktor.util.pipeline.SuspendFunctionGun.resumeRootWith(PipelineContext.kt:211)
	
	20:08:53 [GRAVE]     at io.ktor.util.pipeline.SuspendFunctionGun.loop(PipelineContext.kt:168)
	
	20:08:53 [GRAVE]     at io.ktor.util.pipeline.SuspendFunctionGun.access$loop(PipelineContext.kt:63)
	
	20:08:53 [GRAVE]     at io.ktor.util.pipeline.SuspendFunctionGun$continuation$1.resumeWith(PipelineContext.kt:118)
	
	20:08:53 [GRAVE]     at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:46)
	
	20:08:53 [GRAVE]     at io.ktor.util.pipeline.SuspendFunctionGun.resumeRootWith(PipelineContext.kt:211)
	
	20:08:53 [GRAVE]     at io.ktor.util.pipeline.SuspendFunctionGun.loop(PipelineContext.kt:168)
	
	20:08:53 [GRAVE]     at io.ktor.util.pipeline.SuspendFunctionGun.access$loop(PipelineContext.kt:63)
	
	20:08:53 [GRAVE]     at io.ktor.util.pipeline.SuspendFunctionGun$continuation$1.resumeWith(PipelineContext.kt:118)
	
	20:08:53 [GRAVE]     at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:46)
	
	20:08:53 [GRAVE]     at io.ktor.util.pipeline.SuspendFunctionGun.resumeRootWith(PipelineContext.kt:211)
	
	20:08:53 [GRAVE]     at io.ktor.util.pipeline.SuspendFunctionGun.loop(PipelineContext.kt:168)
	
	20:08:53 [GRAVE]     at io.ktor.util.pipeline.SuspendFunctionGun.access$loop(PipelineContext.kt:63)
	
	20:08:53 [GRAVE]     at io.ktor.util.pipeline.SuspendFunctionGun$continuation$1.resumeWith(PipelineContext.kt:118)
	
	20:08:53 [GRAVE]     at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:46)
	
	20:08:53 [GRAVE]     at kotlinx.coroutines.DispatchedTask.run(Dispatched.kt:238)
	
	20:08:53 [GRAVE]     at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(CoroutineScheduler.kt:594)
	
	20:08:53 [GRAVE]     at kotlinx.coroutines.scheduling.CoroutineScheduler.access$runSafely(CoroutineScheduler.kt:60)
	
	20:08:53 [GRAVE]     at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.kt:742)

**Comment by e5l** [comment]
Hi @hazae41, thanks for the report. I can't reproduce this in tests, could you provide some snippet?

**Comment by hazae41** [comment]
typealias WebSocketHandler = suspend DefaultWebSocketServerSession.() -> Unit

	val webSockets = mutableMapOf<String, WebSocketHandler>()	

	webSockets["/test2"] = {
		send("hello world")
	}

	routing {
		// works
		webSocket("/test1"){
			send("hello world") 
		}

		// does not work
		webSockets.forEach { (path, handler) -> webSocket(path, handler) } 

		// works
		webSockets.forEach{ (path, handler) -> webSocket(path = path, handler = handler) }
		
	}

It seems there is a confusing method overloading for `webSockets()` causing web sockets to not being routed, and no error is thrown.

A solution would be to make the `path` parameter required.

**Comment by penn5** [comment]
I cannot express how happy I am that someone found this bug before me, it would have taken me days rather than hours if this didn't show up in Google.

**Comment by hazae41** [comment]
This bug is one year old and they did not fix it
Maybe you should consider not using this library for production

**Comment by penn5** [comment]
> This bug is one year old and they did not fix it
> Maybe you should consider not using this library for production

I think only the websockets part is bad because it's less used than plain HTTP

**Comment by dmitrievanthony** [comment]
Hi @penn5, @hazae41,

I investigated this issue and figured out that you have got the issue because by default `webSocket(path, handler)` call is resolved into `fun Route.webSocket(protocol: String? = null, handler: suspend DefaultWebSocketServerSession.() -> Unit)` method. I guess it's not what you expected.

To avoid this issue further I deprecated this method with correspondent replacement. This replacement will show how to update your code to avoid the issue. The corresponding PR is on the review so far (https://github.com/ktorio/ktor/pull/1790).
